mlx/backend/metal/device.cpp

```diff
@@ -161,19 +156,32 @@ void CommandEncoder::set_output_array(
   if (concurrent_) {
     concurrent_outputs_.insert(buf);
   } else {
-    outputs_.insert(buf);
+    next_outputs_.insert(buf);
+  }
+}
+
+void CommandEncoder::maybeInsertBarrier() {
+  if (needs_barrier_) {
+    enc_->memoryBarrier(MTL::BarrierScopeBuffers);
+    needs_barrier_ = false;
+    prev_outputs_ = next_outputs_;
```

Comment:
Maybe `std::move`? You could do an `extract()` and move insert loop below but that may be unnecessarily complicated given that we are copying around some integers.